src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,572 @@
+                            return diagnostic;
+                    }
+                }
+
+                return null;
+            }
+
+            static EmbeddedDiagnostic? CheckArray(JsonArrayNode node)
+            {
+                foreach (var child in node.Sequence)
+                {
+                    if (child.Kind == JsonKind.Property)
+                    {
+                        return new EmbeddedDiagnostic(
+                            FeaturesResources.Properties_not_allowed_in_an_array,
+                            ((JsonPropertyNode)child).ColonToken.GetSpan());
+                    }
+                }
+
+                return CheckChildren(node);
+            }
+        }
+
+        private static JsonToken GetFirstToken(JsonNodeOrToken nodeOrToken)
+            => nodeOrToken.IsNode ? GetFirstToken(nodeOrToken.Node.ChildAt(0)) : nodeOrToken.Token;
+
+        private static EmbeddedDiagnostic? GetFirstDiagnostic(JsonNode node)
+        {
+            foreach (var child in node)
+            {
+                var diagnostic = GetFirstDiagnostic(child);
+                if (diagnostic != null)
+                    return diagnostic;
+            }
+
+            return null;
+        }
+
+        private static EmbeddedDiagnostic? GetFirstDiagnostic(JsonNodeOrToken child)
+            => child.IsNode
+                ? GetFirstDiagnostic(child.Node)
+                : GetFirstDiagnostic(child.Token);
+
+        private static EmbeddedDiagnostic? GetFirstDiagnostic(JsonToken token)
+            => GetFirstDiagnostic(token.LeadingTrivia) ?? token.Diagnostics.FirstOrNull() ?? GetFirstDiagnostic(token.TrailingTrivia);
+
+        private static EmbeddedDiagnostic? GetFirstDiagnostic(ImmutableArray<JsonTrivia> list)
+        {
+            foreach (var trivia in list)
+            {
+                var diagnostic = trivia.Diagnostics.FirstOrNull();
+                if (diagnostic != null)
+                    return diagnostic;
+            }
+
+            return null;
+        }
+
+        private ImmutableArray<JsonValueNode> ParseSequence()
+        {
+            try
+            {
+                _recursionDepth++;
+                StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
+                return ParseSequenceWorker();
+            }
+            finally
+            {
+                _recursionDepth--;
+            }
+        }
+
+        private ImmutableArray<JsonValueNode> ParseSequenceWorker()
+        {
+            using var _ = ArrayBuilder<JsonValueNode>.GetInstance(out var result);
+
+            while (ShouldConsumeSequenceElement())
+                result.Add(ParseValue());
+
+            return result.ToImmutable();
+        }
+
+        private JsonSeparatedList ParseCommaSeparatedSequence()
+        {
+            try
+            {
+                _recursionDepth++;
+                StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
```

Comment:
❓I've always wondered why `StackGuard.EnsureSufficientExecutionStack(ref _recursionDepth)` wasn't a thing, where it could return an `IDisposable` instead of having to do try/finally

Reply:
i like that! :)